### PR TITLE
Use setup-python for the free-threading CI job

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by da-woods** [comment]
Windows doesn't seem happy. Unfortunately if it crashes and then times out it doesn't count as an "allowed failure"

**Comment by da-woods** [comment]
I don't think `Py_GIL_DISABLED` should be the problem - I added a check that should spot if you've mismatched that flag and raise an error before it runs anything harmful. It won't work, but it shouldn't give mystery crashes (I hope)

**Comment by da-woods** [comment]
Is there a way for me to download the exact Windows Python version it's using locally?

**Comment by da-woods** [comment]
> Is there a way for me to download the exact Windows Python version it's using locally?

Nevermind. Found it.

**Comment by da-woods** [comment]
Right... the Cython build itself is fixed by the changes in you propose in setup.py.

However, that doesn't force `-DPy_GIL_DISABLED=1` when running the tests. There's a whole bunch of separate commands there that compile Cython modules so it isn't obvious what the best way to set it everywhere is. Maybe just as a CFLAGS in ci-run.sh. Although setting the environmental variable isn't working for me locally on Windows.

Alternatively maybe it'd be easier to edit pyconfig.h as part of the setup?

**Comment by lysnikolaou** [comment]
> They mostly go through distutils/setuptools. There's a suggestion to https://github.com/pypa/setuptools/issues/4662, which is probably better than Cython trying to patch it in for the tests.

Yeah, and there's an [open PR](https://github.com/pypa/distutils/pull/310) for it as well (which unforunately seems to have staled a bit). Maybe it'd be best to add macOS for now and then expand to windows once the fix is merged and there's a new setuptools release with it?

**Comment by da-woods** [comment]
> Maybe it'd be best to add macOS for now and then expand to windows once the fix is merged and there's a new setuptools release with it?

Yes I agree

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -179,9 +180,28 @@ jobs:
...
+            backend: "cpp"
+            allowed_failure: true
+          - os: windows-2019
+            python-version: 3.13t
+            backend: "c"
+            allowed_failure: true
+          - os: windows-2019
+            python-version: 3.13t
+            backend: "cpp"
+            allowed_failure: true
```
Comment (da-woods) [inline .github/workflows/ci.yml:197]:
```suggestion
#          - os: windows-2019
#            python-version: 3.13t
#            backend: "c"
#            allowed_failure: true
#          - os: windows-2019
#            python-version: 3.13t
#            backend: "cpp"
#            allowed_failure: true
```

Comment (da-woods) [reply .github/workflows/ci.yml]: Unfortunately this didn't work too well, and just turned of "allowed_failure" I think

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -51,6 +51,7 @@ jobs:
           - "3.11"
           - "3.12"
           - "3.13"
+          - "3.13t"
         env: [{}]
 
         include:
```
Comment (da-woods) [inline .github/workflows/ci.yml:57]:
```suggestion
        exclude:
          # Windows and free-threading is difficult to build right now
          - os: windows-2019
            python-version: 3.13t
            backend: "c"
          - os: windows-2019
            python-version: 3.13t
            backend: "cpp"
        include:
```

**Pull request merged**